### ISSUE TITLE
Fitting & evaluation aren't working for certain 2D compound models

### ISSUE THREAD

**Issue by jehturner** [astropy/astropy#6071]
When trying to fit 2D compound models (after #6012 was fixed), the model coefficients all end up being unchanged by the fitter, ie. 0.0 for an uninitialized model.

Here is a simplified example, where fitting works with a simple 2D model but the same fitter fails silently on the same data with a completely equivalent compound model:
```python
from astropy.modeling.models import *
from astropy.modeling.fitting import *
import numpy as np

y, x = np.mgrid[0:5, 0:5]
z = x + y

fitter = LevMarLSQFitter()
simple = Chebyshev2D(2,2)
simple_fit = fitter(simple, x, y, z)
print simple_fit(x, y)

# [[  1.23695032e-17   1.00000000e+00   2.00000000e+00   3.00000000e+00
#     4.00000000e+00]
#  [  1.00000000e+00   2.00000000e+00   3.00000000e+00   4.00000000e+00
#     5.00000000e+00]
#  [  2.00000000e+00   3.00000000e+00   4.00000000e+00   5.00000000e+00
#     6.00000000e+00]
#  [  3.00000000e+00   4.00000000e+00   5.00000000e+00   6.00000000e+00
#     7.00000000e+00]
#  [  4.00000000e+00   5.00000000e+00   6.00000000e+00   7.00000000e+00
#     8.00000000e+00]]

fitter = LevMarLSQFitter()
compound = Identity(2) | Chebyshev2D(2,2)
compound_fit = fitter(compound, x, y, z)
print compound_fit(x, y)

# [[ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]]
```
This problem also occurs in the following variations:

  * With other 2D polynomial models, such as Legendre2D.
    - I previously wrote "and Polynomial2D" but that seemed to be working in my last attempt.
  * When starting with Chebyshev coefficients close to the correct values.
  * Using other fitters than LevMarLSQFitter (LinearLSQFitter doesn't work due to #6038).
  * With `Shift() & Shift() | Chebyshev2D(2,2)` (and other combinations) instead of `Identity`.

The problem does not occur with `Const2D`, while the prototypical example, `Gaussian2D`, gives quite different results for the simple & compound cases -- which seems a bit odd, even though it's a completely inappropriate model for these test data.

The equivalent 1D example works normally:
```python
from astropy.modeling.models import *
from astropy.modeling.fitting import *
import numpy as np

x = np.mgrid[0:5]
z = x

fitter = LevMarLSQFitter()
simple = Chebyshev1D(2)
simple_fit = fitter(simple, x, z)
print simple_fit(x)

# [  2.47069853e-17   1.00000000e+00   2.00000000e+00   3.00000000e+00
#    4.00000000e+00]

fitter = LevMarLSQFitter()
compound = Identity(1) | Chebyshev1D(2)
compound_fit = fitter(compound, x, z)
print compound_fit(x)

# [ -1.16880786e-16   1.00000000e+00   2.00000000e+00   3.00000000e+00
#    4.00000000e+00]
```

**Comment by jehturner** [comment]
Update: I've changed the above examples to use a new fitter instance for the compound model, so it's clear that this isn't just something that happens when re-using the existing fitter.

**Comment by jehturner** [comment]
I don't understand the fields well, but the `LevMarLSQFitter.fit_info` for the compound 2D case looks like this, which contains a status message from `scipy.optimize.leastsq`, "The cosine of the angle between func(x) and any column of the Jacobian is at most 0.000000 in absolute value":
```python
{u'cov_x': None, u'fvec': array([ 0., -1., -2., -3., -4., -1., -2., -3., -4., -5., -2., -3., -4.,
       -5., -6., -3., -4., -5., -6., -7., -4., -5., -6., -7., -8.]), u'ierr': 4, u'qtf': array([ 0., -1., -2., -3., -4., -1., -2., -3., -4., -5., -2.]), u'nfev': 12, u'param_cov': None, u'ipvt': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11], dtype=int32), u'param_jac': None, u'message': 'The cosine of the angle between func(x) and any column of the\n  Jacobian is at most 0.000000 in absolute value', u'fjac': array([[-0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0., -0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0., -0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0., -0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0., -0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0., -0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0., -0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0., -0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., -0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., -0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0., -0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.]])}
```

**Comment by jehturner** [comment]
One notable difference between the simple and compound cases is that the compound model has no `fit_deriv` method, but changing the fitter's `estimate_jacobian` parameter doesn't make a difference in either case, which I think rules that out as the origin of the difference.

**Comment by jehturner** [comment]
Comparing the argument values passed by `LevMarLSQFitter` to `scipy.optimize.leastsq` in each 2D case, the only difference is the model instance itself (in `farg[0]`, which `leastsq` passes back to `LevMarLSQFitter.objective_function`).

**Comment by jehturner** [comment]
I think the problem is that when the `parameters` array is varied, the 2D CompoundModel doesn't actually change! Assuming I understand the intended behaviour, it seems like there isn't really a fitting problem here, but the compound models are just broken in some way??

```python
from astropy.modeling.fitting import *
import numpy as np

y, x = np.mgrid[0:5, 0:5]
z = x + y

fitter = LevMarLSQFitter()
simple = Chebyshev2D(1,1)
simple.parameters[:] = [1,1,1,1]
print simple(x, y)

# [[  1.   2.   3.   4.   5.]
#  [  2.   4.   6.   8.  10.]
#  [  3.   6.   9.  12.  15.]
#  [  4.   8.  12.  16.  20.]
#  [  5.  10.  15.  20.  25.]]

fitter = LevMarLSQFitter()
compound = Identity(2) | Chebyshev2D(1,1)
compound.parameters[:] = [1,1,1,1]
print compound(x, y)

# [[ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]]
```
Am I interpreting that correctly?

However, varying the parameters by name does work:
```python
rom astropy.modeling.models import *
from astropy.modeling.fitting import *
import numpy as np

y, x = np.mgrid[0:5, 0:5]
z = x + y

fitter = LevMarLSQFitter()
simple = Chebyshev2D(1,1)
simple.c0_0 = 1.
simple.c1_0 = 1.
simple.c0_1 = 1.
simple.c1_1 = 1.
print simple(x, y)

# [[  1.   2.   3.   4.   5.]
#  [  2.   4.   6.   8.  10.]
#  [  3.   6.   9.  12.  15.]
#  [  4.   8.  12.  16.  20.]
#  [  5.  10.  15.  20.  25.]]

fitter = LevMarLSQFitter()
compound = Identity(2) | Chebyshev2D(1,1)
compound.c0_0_1 = 1.
compound.c1_0_1 = 1.
compound.c0_1_1 = 1.
compound.c1_1_1 = 1.
print compound(x, y)

# [[  1.   2.   3.   4.   5.]
#  [  2.   4.   6.   8.  10.]
#  [  3.   6.   9.  12.  15.]
#  [  4.   8.  12.  16.  20.]
#  [  5.  10.  15.  20.  25.]]
```

**Comment by jehturner** [comment]
The underlying `Chebyshev2D` model actually *does* change as expected in response to setting the compound model's `.parameters[:]` as above and evaluates properly ... it's just the compound model itself that doesn't evaluate correctly unless the parameters are set using their named attributes.

```python
from astropy.modeling.models import *
from astropy.modeling.fitting import *
import numpy as np

y, x = np.mgrid[0:5, 0:5]
z = x + y

fitter = LevMarLSQFitter()
compound = Identity(2) | Chebyshev2D(1,1)
compound.parameters[:] = [1,1,1,1]
print compound(x, y)

# [[ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]
#  [ 0.  0.  0.  0.  0.]]

print compound[1](x, y)

# [[  1.   2.   3.   4.   5.]
#  [  2.   4.   6.   8.  10.]
#  [  3.   6.   9.  12.  15.]
#  [  4.   8.  12.  16.  20.]
#  [  5.  10.  15.  20.  25.]]
```
It seems like `ExpressionTree.evaluate()` is producing a broken `_evaluate()` method in `CompoundModelMeta`.

**Comment by nden** [comment]
@jehturner Just a quick note that I looked at some of these issues and have a better idea (still incomplete) of what needs to be done to resolve this. Most of the issues are related to constructing the derivative incorrectly or not constructing it at all. Choosing a fitting algorithm which does not use a derivative may help, e.g. `SimplexLSQFitter` or `SLSQPLSQFitter`. 
Unfortunately I don't have time to work on this for v2.0.

**Comment by jehturner** [comment]
The new result after the changes in PR #6085 matches that from the equivalent simple fit:
```python
from astropy.modeling.models import *
from astropy.modeling.fitting import *
import numpy as np

y, x = np.mgrid[0:5, 0:5]
z = x + y

fitter = LevMarLSQFitter()
simple = Chebyshev2D(1,1)
simple.parameters[:] = [1,1,1,1]
print simple(x, y)

# [[  1.   2.   3.   4.   5.]
#  [  2.   4.   6.   8.  10.]
#  [  3.   6.   9.  12.  15.]
#  [  4.   8.  12.  16.  20.]
#  [  5.  10.  15.  20.  25.]]

fitter = LevMarLSQFitter()
compound = Identity(2) | Chebyshev2D(1,1)
compound.parameters[:] = [1,1,1,1]
print compound(x, y)

# [[  1.   2.   3.   4.   5.]
#  [  2.   4.   6.   8.  10.]
#  [  3.   6.   9.  12.  15.]
#  [  4.   8.  12.  16.  20.]
#  [  5.  10.  15.  20.  25.]]
```
I'll add a test and changelog entry if you're happy with this.

**Comment by jehturner** [comment]
@nden, thanks for the info. I'll also need to return to my experiment for the time being, where I can work around some of these issues. Maybe you want to change the milestone here, though presumably it will be moot once we close this.

**Comment by pllim** [comment]
Fixed in #6085.
